File: poisson_recon/Module/poisson_reconstructor.py
import os
import numpy as np
import open3d as o3d
from typing import Union

from poisson_recon.Data.poisson_params import PoissonParams
from poisson_recon.Method.cmd import runCMD
from poisson_recon.Method.time import getCurrentTime
from poisson_recon.Method.path import createFileFolder, removeFile, renameFile


class PoissonReconstructor(object):
    def __init__(self, poisson_params: PoissonParams = PoissonParams()) -> None:
        self.poisson_params = poisson_params

        self.loaded_mesh_file_path = ""
        self.loaded_vertices_list = []
        self.loaded_faces_list = []
        return

    def updateParams(self,
                     degree: int=1,
                     bType: int=3,
                     depth: int=8,
                     scale: float=1.1,
                     samplesPerNode: float=1.5,
                     pointWeight: Union[float, None]=None,
                     iters: int=8,
                     confidence: float=0,
                     confidenceBias: float=0,
                     primalGrid: bool=False,
                     linearFit: bool=False,
                     polygonMesh: bool=False) -> bool:
        self.poisson_params.degree = degree
        self.poisson_params.bType = bType
        self.poisson_params.depth = depth
        self.poisson_params.scale = scale
        self.poisson_params.samplesPerNode = samplesPerNode
        self.poisson_params.pointWeight = pointWeight
        self.poisson_params.iters = iters
        self.poisson_params.confidence = confidence
        self.poisson_params.confidenceBias = confidenceBias
        self.poisson_params.primalGrid = primalGrid
        self.poisson_params.linearFit = linearFit
        self.poisson_params.polygonMesh = polygonMesh
        return True

    def estimateNormal(self, pcd_file_path: str, save_pcd_file_path: str) -> bool:
        if not os.path.exists(pcd_file_path):
            print("[ERROR][PoissonReconstructor::reconMeshFile]")
            print("\t pcd file not exist!")
            print("\t pcd_file_path:", pcd_file_path)
            return False

        createFileFolder(save_pcd_file_path)

        pcd = o3d.io.read_point_cloud(pcd_file_path)

        if not pcd.has_normals():
            pcd.estimate_normals()

        pcd.normalize_normals()
        o3d.io.write_point_cloud(save_pcd_file_path, pcd)

        return True

    def reconMeshFile(
        self,
        pcd_file_path: str,
        save_mesh_file_path: str,
        overwrite: bool = False,
        print_progress: bool = False,
    ) -> Union[str, None]:
        if not os.path.exists(pcd_file_path):
            print("[ERROR][PoissonReconstructor::reconMeshFile]")
            print("\t pcd file not exist!")
            print("\t pcd_file_path:", pcd_file_path)
            return None

        normal_pcd_file_path = "./output/normal_pcd.ply"

        self.estimateNormal(pcd_file_path, normal_pcd_file_path)

        full_save_mesh_file_path = (
            save_mesh_file_path[:-4] + self.poisson_params.toLogStr() + ".ply"
        )

        if os.path.exists(full_save_mesh_file_path):
            if overwrite:
                removeFile(full_save_mesh_file_path)
            else:
                print("[ERROR][PoissonReconstructor::reconMeshFile]")
                print("\t full save mesh file already exist!")
                print("\t full_save_mesh_file_path:", full_save_mesh_file_path)
                return None

        createFileFolder(save_mesh_file_path)

        tmp_save_mesh_file_path = full_save_mesh_file_path[:-4] + "_tmp.ply"

        cmd = (
            "../PoissonRecon/Bin/Linux/PoissonRecon"
            + " --in "
            + normal_pcd_file_path
            + " --out "
            + tmp_save_mesh_file_path
            + self.poisson_params.toCMDStr()
        )

        if print_progress:
            print("[INFO][PoissonReconstructor::reconMeshFile]")
            print("\t start PoissonRecon...")
            print("\t cmd:")
            print(cmd)
        if not runCMD(cmd):
            print("[ERROR][PoissonReconstructor::reconMeshFile]")
            print("\t runCMD failed!")
            print("\t cmd:", cmd)
            return None

        if not os.path.exists(tmp_save_mesh_file_path):
            print("[ERROR][PoissonReconstructor::reconMeshFile]")
            print("\t mesh file save failed!")
            print("\t save_mesh_file_path:", save_mesh_file_path)
            return None

        renameFile(tmp_save_mesh_file_path, full_save_mesh_file_path)
        return full_save_mesh_file_path

    def reconMeshFileFromPoints(self,
                                points: Union[np.ndarray, list],
                                save_mesh_file_path: str,
                                overwrite: bool=False,
                                print_progress: bool = False) -> Union[str, None]:
        if isinstance(points, list):
            points = np.array(points)

        is_flatten = len(points.shape) == 1
        if is_flatten:
            points = points.reshape(-1, 3)

        tmp_pcd_file_path = './output/tmp_pcd_generated_by_poisson_recon.ply'
        createFileFolder(tmp_pcd_file_path)
        
        pcd = o3d.geometry.PointCloud()
        pcd.points = o3d.utility.Vector3dVector(points)
        o3d.io.write_point_cloud(tmp_pcd_file_path, pcd, write_ascii=True)

        return self.reconMeshFile(tmp_pcd_file_path, save_mesh_file_path, overwrite, print_progress)


    def autoReconMeshFileFromPointsList(self,
                                        points: list,
                                        save_mesh_folder_path: str,
                                        overwrite: bool=False,
                                        print_progress: bool=False) -> str:
        if save_mesh_folder_path[-1] != '/':
            save_mesh_folder_path += '/'

        save_mesh_file_path = save_mesh_folder_path + getCurrentTime() + '_poisson_recon.ply'

        final_save_mesh_file_path = self.reconMeshFileFromPoints(points,
                                                                 save_mesh_file_path,
                                                                 overwrite,
                                                                 print_progress)
        if final_save_mesh_file_path is None:
            return ""

        return final_save_mesh_file_path

    def loadReconMeshFile(self, mesh_file_path: str) -> bool:
        if self.loaded_mesh_file_path == mesh_file_path:
            return True

        self.loaded_mesh_file_path = ""
        self.loaded_vertices_list = []
        self.loaded_faces_list = []

        if not os.path.exists(mesh_file_path):
            print('[ERROR][PoissonReconstructor::loadReconMeshFile]')
            print('\t mesh file not exist!')
            return False

        mesh = o3d.io.read_triangle_mesh(mesh_file_path)
        self.loaded_vertices_list = np.asarray(mesh.vertices).reshape(-1).tolist()
        self.loaded_faces_list = np.asarray(mesh.triangles).reshape(-1).tolist()
        return True

    def getLoadedVerticesList(self) -> list:
        return self.loaded_vertices_list

    def getLoadedFacesList(self) -> list:
        return self.loaded_faces_list

File: src/poisson_reconstructor.cpp
#include "poisson_reconstructor.h"
#include "cmd.h"
#include "normal.h"
#include <filesystem>
#include <iostream>
#include <string>

const bool PoissonReconstructor::isValid() {
  if (poisson_recon_bin_file_path_ == "") {
    return false;
  }

  return true;
}

const bool PoissonReconstructor::setPoissonReconFolderPath(
    const std::string &poisson_recon_folder_path) {
  poisson_recon_bin_file_path_ = "";

  if (!std::filesystem::exists(poisson_recon_folder_path)) {
    std::cout << "[ERROR][PoissonReconstructor::setPoissonReconFolderPath]"
              << std::endl;
    std::cout << "\t poisson recon bin folder not exist!" << std::endl;
    std::cout << "\t poisson_recon_folder_path: " << poisson_recon_folder_path
              << std::endl;

    return false;
  }

  const std::string poisson_recon_bin_file_path_linux =
      poisson_recon_folder_path + "Bin/Linux/PoissonRecon";

  if (std::filesystem::exists(poisson_recon_bin_file_path_linux)) {
    poisson_recon_bin_file_path_ = poisson_recon_bin_file_path_linux;

    return true;
  }

  std::cout << "[ERROR][PoissonReconstructor::setPoissonReconFolderPath]"
            << std::endl;
  std::cout << "\t poisson recon bin file for linux not exist!" << std::endl;
  std::cout << "\t poisson_recon_bin_file_path_linux: "
            << poisson_recon_bin_file_path_linux << std::endl;

  return false;
}

const bool PoissonReconstructor::updateParams(
    const int &degree, const int &bType, const int &depth, const float &scale,
    const float &samplesPerNode, const float &pointWeight, const int &iters,
    const float &confidence, const float &confidenceBias,
    const bool &primalGrid, const bool &linearFit, const bool &polygonMesh) {
  poisson_params_.degree = degree;
  poisson_params_.bType = bType;
  poisson_params_.depth = depth;
  poisson_params_.scale = scale;
  poisson_params_.samplesPerNode = samplesPerNode;
  poisson_params_.pointWeight = pointWeight;
  poisson_params_.iters = iters;
  poisson_params_.confidence = confidence;
  poisson_params_.confidenceBias = confidenceBias;
  poisson_params_.primalGrid = primalGrid;
  poisson_params_.linearFit = linearFit;
  poisson_params_.polygonMesh = polygonMesh;
  return true;
}

const bool
PoissonReconstructor::reconMeshFile(const std::string &pcd_file_path,
                                    const std::string &save_mesh_file_path,
                                    const bool &overwrite) {
  if (!std::filesystem::exists(pcd_file_path)) {
    std::cout << "[ERROR][PoissonReconstructor::reconMeshFile]" << std::endl;
    std::cout << "\t pcd file not exist!" << std::endl;
    std::cout << "\t pcd_file_path: " << pcd_file_path << std::endl;

    return false;
  }

  if (std::filesystem::exists(save_mesh_file_path)) {
    if (!overwrite) {
      std::cout << "[ERROR][PoissonReconstructor::reconMeshFile]" << std::endl;
      std::cout << "\t save mesh file already exist!" << std::endl;
      std::cout << "\t save_mesh_file_path: " << save_mesh_file_path
                << std::endl;

      return false;
    } else {
      std::filesystem::remove(save_mesh_file_path);
    }
  }

  const std::string normal_pcd_file_path = "./output/normal_pcd.ply";

  if (!estimateNormal(pcd_file_path, normal_pcd_file_path, true)) {
    std::cout << "[ERROR][PoissonReconstructor::reconMeshFile]" << std::endl;
    std::cout << "\t estimateNormal failed!" << std::endl;

    return false;
  }

  const std::string full_save_mesh_file_basepath =
      save_mesh_file_path.substr(0, save_mesh_file_path.length() - 4) +
      poisson_params_.toLogStr();

  std::string full_save_mesh_file_path;

  int save_mesh_idx = 0;

  while (true) {
    full_save_mesh_file_path = full_save_mesh_file_basepath + "_" +
                               std::to_string(save_mesh_idx) + ".ply";

    if (std::filesystem::exists(full_save_mesh_file_path)) {
      ++save_mesh_idx;
      continue;
    }

    break;
  }

  const std::string tmp_save_mesh_file_path =
      full_save_mesh_file_path.substr(0,
                                      full_save_mesh_file_path.length() - 4) +
      "_tmp.ply";

  if (std::filesystem::exists(tmp_save_mesh_file_path)) {
    std::filesystem::remove(tmp_save_mesh_file_path);
  }

  std::string command = poisson_recon_bin_file_path_ + " --in " +
                        normal_pcd_file_path + " --out " +
                        tmp_save_mesh_file_path + poisson_params_.toCMDStr();

  std::cout << "[INFO][PoissonReconstructor::reconMeshFile]" << std::endl;
  std::cout << "\t start recon mesh from points by PoissonRecon..."
            << std::endl;
  if (!runCMD(command)) {
    std::cout << "[ERROR][PoissonReconstructor::reconMeshFile]" << std::endl;
    std::cout << "\t runCMD failed!" << std::endl;

    return false;
  }

  if (!std::filesystem::exists(tmp_save_mesh_file_path)) {
    std::cout << "[ERROR][PoissonReconstructor::reconMeshFile]" << std::endl;
    std::cout << "\t mesh file save failed!" << std::endl;
    std::cout << "\t save_mesh_file_path: " << tmp_save_mesh_file_path
              << std::endl;

    return false;
  }

  try {
    std::filesystem::rename(tmp_save_mesh_file_path, full_save_mesh_file_path);
  } catch (const std::filesystem::filesystem_error &e) {
    std::cout << "[ERROR][PoissonReconstructor::reconMeshFile]" << std::endl;
    std::cout << "\t rename mesh file failed!" << std::endl;
    std::cout << "\t error: " << e.what() << std::endl;

    return false;
  }

  saved_mesh_file_path_vec_.emplace_back(full_save_mesh_file_path);

  return true;
}

const int PoissonReconstructor::getSavedMeshNum() {
  return saved_mesh_file_path_vec_.size();
}

const int PoissonReconstructor::toValidMeshIdxs(const int &mesh_idx) {
  int valid_mesh_idx = mesh_idx;

  if (valid_mesh_idx < 0) {
    valid_mesh_idx += saved_mesh_file_path_vec_.size();
  }

  if (valid_mesh_idx >= saved_mesh_file_path_vec_.size()) {
    std::cout << "given mesh idx [" << valid_mesh_idx << "] out of range!"
              << std::endl;
    std::cout << "the valid range is [0 - "
              << saved_mesh_file_path_vec_.size() - 1 << "]" << std::endl;

    return -1;
  }

  return valid_mesh_idx;
}
